### PR TITLE
feat(jans-cedarling): cap .cjar archive sizes to prevent zip-bomb resource exhaustion

### LINKED ISSUE #14896: feat(jans-cedarling): cap .cjar archive sizes to prevent zip-bomb resource exhaustion
## Problem

`.cjar` policy stores are ZIP archives. `ArchiveVfs` validates path traversal
and ZIP integrity, but reads every entry with an unbounded `read_to_end`
(`archive_handler.rs:324`). There is no limit on:

- decompressed size of a single entry (zip-bomb: a tiny archive can expand to a
  huge in-memory file)
- total archive size
- number of entries (also makes `read_dir` / `is_directory_locked` O(n) per call)

This is the same class of issue already fixed for status lists in
`StatusList::parse` (10 MB decompressed cap), but not yet for policy stores. #14821 

## Scope of limits

Two distinct sizes, two limits:

- **Downloaded (compressed) archive size** — already capped by
  `CEDARLING_HTTP_MAX_RESPONSE_SIZE_BYTES` via `read_response_capped`.
- **Decompressed per-entry size** — currently uncapped, covered by the new
  property below.

## Proposed change

Enforce limits during `ArchiveVfs::from_reader` validation and in `read_file`:

- per-entry decompressed size
- total archive size (also covers `CjarFile` / `ArchiveBytes`, which the HTTP
  cap does not)
- maximum entry count

Exceeded limits return a typed `ArchiveError` variant (e.g.
`EntrySizeExceeded`, `ArchiveSizeExceeded`, `TooManyEntries`) instead of OOMing.

## Bootstrap property

- `CEDARLING_POLICY_STORE_MAX_FILE_SIZE` (default 10 MB) — cap on the
  decompressed size of a single archive entry.

When `CEDARLING_HTTP_MAX_RESPONSE_SIZE_BYTES` is not set explicitly, it must
fall back to this value rather than using its own independent default, so a
download is never larger than the largest entry we are willing to decompress.

Update `default_config.yaml` and
`docs/cedarling/reference/cedarling-properties.md` in the same change.

## Acceptance criteria

- A `.cjar` with a zip-bomb entry fails with a typed error, not OOM.
- Unset `CEDARLING_HTTP_MAX_RESPONSE_SIZE_BYTES` falls back to
  `CEDARLING_POLICY_STORE_MAX_FILE_SIZE`.
- The property is honored from env/JSON/YAML and documented.
- Tests cover: oversized entry, oversized archive, too many entries, and the
  exact-at-limit boundary case.

### PR DESCRIPTION
**Pull request by haileyesus2433**
### Prepare

- [x] Read [PR guidelines](https://github.com/JanssenProject/jans/blob/main/docs/CONTRIBUTING.md#prs)
- [x] Read [license information](https://github.com/JanssenProject/jans/blob/main/LICENSE)

-------------------

#### Target issue
  
  
  
  
closes #14896

#### Implementation Details
  

-------------------
### Test and Document the changes
- [x] Static code analysis has been run locally and issues have been fixed
- [x] Relevant unit and integration tests have been added/updated
- [x] Relevant documentation has been updated if any (i.e. user guides, installation and configuration guides, technical design docs etc)

Please check the below before submitting your PR. The PR will not be merged if there are no commits that start with `docs:` to indicate documentation changes or if the below checklist is not selected.
- [ ] **I confirm that there is no impact on the docs due to the code changes in this PR.**

## Summary by CodeRabbit

* **New Features**
  * Added configurable safety limits for Cedar Archive files, including per-entry size, total decompressed size, and entry count.
  * Added `CEDARLING_POLICY_STORE_MAX_FILE_SIZE`, defaulting to 10 MB. Total archive size is capped at 10 times this value, with a 10,000-entry maximum; setting it to `0` disables size caps.
  * Documented that policy-store archive limits and HTTP response limits are configured separately.

* **Bug Fixes**
  * Prevented oversized or potentially harmful archives from being loaded or refreshed.
  * Clarified HTTP response size-limit error messages.

### CODE REVIEW

**Comment by mo-auto** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/janssenproject/pr-checks/c6291092-5313-40fc-8013-e3cb44589fdb?refs=8c1c7ab54df7d0d0e930ec0ac4b4d955f6e087a2&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: jans-cedarling/cedarling/src/tests/policy_store_loader.rs [hunk 1]
```diff
@@ -1222,8 +1222,53 @@ async fn test_cjar_url_handles_http_error() {
 ///
 /// This tests the `load_policy_store_archive_bytes` function which is the
 /// underlying mechanism used by `CjarUrl` and is WASM-compatible.
+/// `CEDARLING_POLICY_STORE_MAX_FILE_SIZE` must reach the archive loader, not
+/// just the `ArchiveVfs::from_buffer` call site. A store that loads fine at the
+/// default cap must be rejected once the configured cap drops below it.
+#[test]
+async fn test_configured_max_file_size_reaches_archive_loader() {
```
[the remark this answers is not in the document]
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/tests/policy_store_loader.rs:1229]: cedarling/src/tests/policy_store_loader.rs:33 is use tokio::test;, which shadows the std attribute. Every async test in that file uses plain #[test], and this runs and passes.

File: jans-cedarling/cedarling/src/bootstrap_config/decode.rs [hunk 2]
```diff
@@ -234,3 +243,131 @@ fn resolve_log_type(
...
+    use super::*;
+    use crate::common::policy_store::archive_handler::ArchiveLimits;
+
+    /// Minimal config body; a policy store source is required, and the JSON
+    /// string source keeps these tests off the filesystem and network.
+    fn raw_config_json(extra: &str) -> String {
+        format!(
+            r#"{{
+                "CEDARLING_APPLICATION_NAME": "test",
+                "CEDARLING_POLICY_STORE_LOCAL": "{{\"cedar_version\":\"v4.0.0\",\"policy_stores\":{{}}}}"
```
[the remark this answers is not in the document]
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/bootstrap_config/decode.rs]: inline JSON is now rejected as a legacy store, so I converted it to YAML

File: jans-cedarling/cedarling/src/common/policy_store/archive_handler.rs [hunk 3]
```diff
@@ -59,20 +105,33 @@ where
...
         })?;
 
+        // Checked before the loop: the count is a per-call cost multiplier for
+        // `read_dir` / `is_directory_locked`, not just a memory concern.
+        if limits.max_entries > 0 && archive.len() > limits.max_entries {
+            return Err(ArchiveError::TooManyEntries {
+                count: archive.len(),
+                limit: limits.max_entries,
+            });
+        }
```
[the remark this answers is not in the document]
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/common/policy_store/archive_handler.rs:137]: The point is fair: ZipArchive::new parses the full central directory before we can read archive.len(). But downloads are already bounded by the HTTP cap (~227k entries at 10 MB, given the 46-byte minimum central-directory record), and the remaining exposure is operator-supplied local .cjar files. Replacing the zip crate's parser with our own to close that would add far more risk than it removes.

File: jans-cedarling/cedarling/src/bootstrap_config/raw_config/config.rs [hunk 4]
```diff
@@ -552,6 +566,7 @@ fn get_cedarling_env_vars() -> HashMap<String, serde_json::Value> {
 #[cfg(test)]
 mod tests {
     use super::*;
+    use crate::common::policy_store::archive_handler::ArchiveLimits;
```
[the remark this answers is not in the document]
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/bootstrap_config/raw_config/config.rs:569]: this is fine since The block's existing order is super::*, then crate::, then std::, then test_utils  already not std-first.

File: jans-cedarling/cedarling/src/common/policy_store/archive_handler.rs [hunk 5]
```diff
@@ -630,3 +782,257 @@ mod tests {
...
+            max_total_size: SMALL_LIMIT + 1000,
+            max_entries: 100,
+        };
+        let vfs = ArchiveVfs::from_buffer(bytes, limits)
+            .expect("understated sizes must pass the construction-time total");
+
+        assert_eq!(
+            vfs.read_file("a.json").expect("first entry fits").len(),
+            entry
+        );
```
[the remark this answers is not in the document]
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/common/policy_store/archive_handler.rs:984]: this is fine

File: jans-cedarling/cedarling/src/bootstrap_config/policy_store_config.rs [hunk 6]
```diff
@@ -184,6 +205,7 @@ impl TryFrom<PolicyStoreConfigRaw> for PolicyStoreConfig {
         Ok(Self {
             source,
             refresh_interval_secs: 0,
```
Comment (olehbozhok) [inline jans-cedarling/cedarling/src/bootstrap_config/policy_store_config.rs:207]: Why this is always zero?

**Review by olehbozhok — COMMENTED** [review]
## `CEDARLING_HTTP_MAX_RESPONSE_SIZE_BYTES` falling back to `CEDARLING_POLICY_STORE_MAX_FILE_SIZE` couples two unrelated limits

`cedarling/src/bootstrap_config/decode.rs`

```rust
max_response_size_bytes: match raw
    .http_client_max_response_size_bytes
    .unwrap_or(raw.policy_store_max_file_size)
{
    0 => None,
    n => Some(n),
},
```

These two properties guard different things:

- `CEDARLING_HTTP_MAX_RESPONSE_SIZE_BYTES` bounds **every** HTTP response body Cedarling buffers — JWKS, OIDC discovery, status list, Lock Server, policy store.
- `CEDARLING_POLICY_STORE_MAX_FILE_SIZE` bounds how far a single `.cjar` entry may **decompress**.

After this change one knob drives both, and the knob is named after only one of them. An operator tuning archive decompression silently moves the limit on network fetches that have nothing to do with the policy store.

### Why this is a problem

**1. `CEDARLING_POLICY_STORE_MAX_FILE_SIZE=0` disables the HTTP cap globally.**

`0` is the documented "no cap" sentinel for the archive limit. Through `unwrap_or` it reaches the `0 => None` arm, so setting it removes the response-size cap from JWKS, OIDC discovery, status list and Lock Server requests as well. An operator who wanted "don't limit how large my legitimate policy store may decompress" gets "let any upstream stream an unbounded body into memory" — the exact exhaustion vector the HTTP cap exists to close. Nothing in the property name or in its documented meaning signals this. This is covered by the `zero_disables_each_cap_independently` test, so it is intentional, but it is worth reconsidering: it turns a policy-store tuning knob into a security control for unrelated subsystems.

**2. A small archive cap silently shrinks unrelated downloads.**

Symmetrically, setting `CEDARLING_POLICY_STORE_MAX_FILE_SIZE=65536` because the policy store is small now caps *every* HTTP response at 64 KB. A status list or a large JWKS then fails with a response-size error that gives no hint that a policy-store property caused it.

**3. The stated justification does not hold.**

The comment says the fallback exists "so a download is never larger than the largest entry we would decompress". But `ArchiveLimits::from_max_file_size` derives `max_total_size = 10 * max_entry_size`, so an archive may legitimately decompress to ten times the per-entry cap. The download limit therefore ends up 10x stricter than what the archive limits themselves permit, while the comment describes a guarantee the code does not provide. If the coupling is kept, the matching quantity is `max_total_size`, not `max_entry_size`.

### Suggested resolution

Either option is fine; both remove the surprising direction of the coupling.

**A. Decouple.** Give `CEDARLING_HTTP_MAX_RESPONSE_SIZE_BYTES` its own default again (`HttpClientConfig::DEFAULT_MAX_RESPONSE_SIZE_BYTES`) and drop the fallback. The field goes back to `u64`, the `Option` handling in `decode.rs` disappears, and the `BootstrapConfigRaw` field type stops being a breaking change for downstream Rust consumers. Cost: serving a large `.cjar` over HTTP requires raising both properties explicitly.

**B. Let the fallback only raise the limit, never lower it, and stop propagating `0`.**

```rust
max_response_size_bytes: match raw.http_client_max_response_size_bytes {
    // Explicit value wins, including the documented `0` opt-out.
    Some(0) => None,
    Some(n) => Some(n),
    // Unset: a large archive cap may raise the download limit, but a small
    // or disabled one must not shrink it — JWKS / OIDC / status list / Lock
    // fetches are not collateral of a policy-store knob.
    None => {
        let archive_total =
            ArchiveLimits::from_max_file_size(raw.policy_store_max_file_size).max_total_size;
        Some(archive_total.max(HttpClientConfig::DEFAULT_MAX_RESPONSE_SIZE_BYTES))
    },
},
```

Resulting behaviour:

| Configuration | Current | With B |
| --- | --- | --- |
| neither set | 10 MB | 100 MB (`max_total_size`) |
| `MAX_FILE_SIZE=0` | **no HTTP cap at all** | 10 MB (HTTP default) |
| `MAX_FILE_SIZE=64KB` | **all HTTP responses capped at 64 KB** | 10 MB |
| `MAX_FILE_SIZE=50MB` | 50 MB | 500 MB |
| `HTTP_MAX_RESPONSE_SIZE_BYTES=0` | no cap | no cap |

**C. The cleanest variant, if the plumbing is acceptable.** Apply the archive-derived limit per request instead of globally. `HttpClient::read_response_capped` already takes the cap as a parameter internally, so an override variant would only need to be threaded through `load_policy_store_from_uri`, `load_policy_store_from_cjar_url` and the refresh worker tick. Policy-store fetches then get the archive-derived cap and every other endpoint keeps the HTTP default, with no global coupling at all.

Whichever option is taken, the corresponding paragraphs in `docs/cedarling/reference/cedarling-properties.md` need rewording — the current text describes the "never larger than the largest entry we would decompress" guarantee discussed in point 3.

File: jans-cedarling/cedarling/src/bootstrap_config/policy_store_config.rs (hunk 6 above)
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/bootstrap_config/policy_store_config.rs:207]: It's zero because PolicyStoreConfigRaw only carries source and path, so there's no refresh interval to read  0 is the documented "refresh disabled" default.

**Review by dagregi — DISMISSED** [review]

File: jans-cedarling/cedarling/src/http_utils/mod.rs [hunk 7]
```diff
@@ -119,9 +119,12 @@ pub enum HttpRequestReasonError {
     InvalidUtf8(#[source] std::string::FromUtf8Error),
     #[error("failed to read response body bytes: {0}")]
     DecodeResponseBytes(#[source] reqwest::Error),
+    // Names the property so an operator can tell which knob to raise, whichever
+    // endpoint (JWKS, status list, policy store, ...) tripped it.
     #[error(
         "response body exceeds the configured limit of {limit} bytes \
-         (read {read_so_far} bytes before stopping)"
+         (read {read_so_far} bytes before stopping); raise \
+         CEDARLING_HTTP_MAX_RESPONSE_SIZE_BYTES to allow larger responses"
```
[the remark this answers is not in the document]
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/http_utils/mod.rs]: resolved in https://github.com/JanssenProject/jans/pull/15050/commits/b621cff32af246b79e7a5fa1d6065e3f926e512c

**Comment by haileyesus2433** [comment]
@coderabbitai approve

**Comment by olehbozhok** [comment]
@coderabbitai approve

**Review by olehbozhok — APPROVED** [review]
